oc_forward_df and oc_reverse_df functions

jessesadler opened the pull request:
This pull request makes what I think is a fairly minor change to `oc_parse` to use `flatten = true` instead of `simplifyVector = FALSE`, as I noted in #40. This makes creating a data frame from the results really easy, so as long as @dpprdan does not have any objections to the change to the output of `oc_forward` and `oc_reverse`, I think this provides a good base to move forward. I also do not think that the results of `oc_forward` and `oc_reverse`, are significantly modified if you want to have the list be returned.

I also created `oc_forward_df` and `oc_reverse_df` functions that are built on @dpprdan's `oc_forward` and `oc_reverse`. They use a helper function of `oc_parse_df`, which extracts the results data frame and tidies up column names. There is no vectorization yet, but I think this is a good start.

Commit:
Change fromJSON to use flatten

Commit:
Merge branch 'devel' of https://github.com/ropensci/opencage into devel

Commit:
oc_forward_df and oc_reverse_df

Commit:
Merge branch 'devel' of https://github.com/ropensci/opencage into devel

Commit:
Replace lapply chain in opencage_format to work with flatten=true

Commit:
Vectorize oc_forward_df and oc_reverse_df

dpprdan commented:
Just quickly: `flatten = true` and `simplify* = TRUE` unfortunately changes the output of `oc_forward` and `oc_reverse` significantly. E.g. the `results` list becomes a data.frame, so the output is relatively similar to `opencage_forward`. Which also means that it breaks `opencage_forward`. So I have been looking for a way to avoid the `flatten = true` etc. but without real success so far. Giving up for now....

jessesadler commented:
I realize that `flatten = TRUE` messes up the `opencage_format`, but I think we should just take out the `lapply` chain and use the data frame from `flatten =TRUE` as I noted in #43. I have a fix for this in my local copy. I will add it to the pull request, so that everyone can see how it changes the output for `opencage_format`.

jessesadler commented:
PR now includes changes to `opencage_format` to work with `flatten = TRUE` as discussed above and in #43. I also used @dpprdan's vectorization of `oc_forward` and `oc_reverse` to vectorize `oc_forward_df` and `oc_reverse_df`. They do not work with data frames as input or with queries that do not return a response, but I am working on these issues.

This pull request changes `oc_forward` and `oc_reverse` to use `flatten = TRUE` as we have discussed, but otherwise leaves it unchanged. `opencage_forward` and `opencage_reverse` work properly and the "results" are in a data frame more aligned to the parsing of the API. And `oc_forward_df` and `oc_reverse_df` both work with either a single query or vector of queries. If these changes are acceptable, this would close #40 and #43.

Commit:
Merge branch 'devel' of https://github.com/ropensci/opencage into parse_df

Commit:
Merge branch 'devel' of https://github.com/ropensci/opencage into devel

Commit:
Deal with null results and add query column to df format

Commit:
Data frame or vector input for oc_forward_df

Commit:
tibbles for output of oc_parse_df

Commit:
output options for oc_forward_df

dpprdan commented on R/reverse_geocoding.R:
```diff
@@ -77,6 +77,40 @@ oc_reverse <-
+           bounds = NULL,
+           countrycode = NULL,
+           language = NULL,
+           limit = 10,
```
Maybe set the default to `limit = 1`?

dpprdan commented on R/forward_geocoding.R:
```diff
@@ -69,6 +69,38 @@ oc_forward <-
+           bounds = NULL,
+           countrycode = NULL,
+           language = NULL,
+           limit = 10,
```
Maybe set the default to `limit = 1`? If you geocode multiple addresses in a dataframe you usually want to have one result per address, I think. Said differently, it may be confusing to have multiple results per source address.

jessesadler replied:
I am totally fine with using default of 1 result with data frame output.

dpprdan commented on R/forward_geocoding.R:
```diff
@@ -69,6 +69,38 @@ oc_forward <-
+           no_dedupe = FALSE,
+           no_record = FALSE,
+           abbrv = FALSE,
+           add_request = TRUE) {
```
The request is missing in the output. That is probably not a problem though, if we `bind_cols` the  source_df and the results_df (as discussed [here](https://github.com/ropensci/opencage/pull/31#issuecomment-365442732).  We'd have to make sure though, that the source_df and the results_df are of equal length, of course. Otherwise we would have to join, which can get messy if there are duplicate addresses.

jessesadler replied:
I have worked beyond this pull request and created an output option that uses `bind_cols` and forces the limit to 1, because I think that has to be the most useful output. This gives output options for either returning multiple results with a query column or one result and combining the input data frame and the results data frame.

dpprdan commented:
Do you already have something for "data frames as input"?

```r
oc_forward_df <- function(data, col, ...){
  col_qtd <- rlang::enquo(col)
  placename <- data[[rlang::quo_text(col_qtd)]]
}
```

The dots are not meant literally but as "all other options". `placename` should then be a character vector and can be fed into `oc_forward` and so forth. Do we also want to allow a character vector as input? Then this would have to become a bit more complicated. 

I'm curious how you want to allow "output options for either returning multiple results with a query column" in addition to "combining the input data frame and the results data frame". Can you already push your new commits to your repo (to update the PR)?

Commit:
oc_reverse_df data frame input and output options

Commit:
Merge branch 'parse_df' into devel

jessesadler commented:
@dpprdan I just added to the pull request. The changes do a couple of things:

- Add query column to the results and deal with instances when no results are returned
- Enable both `oc_forward_df` and `oc_reverse_df` to take either vectors or data frames as input. This involves some work arounds, so you may want to check out how I do it, but it has been working in my tests. I used `substitute` to quote the column of the data frame, but we could easily use `rlang` if we want. The way that the two functions deal with vector vs data frame input is a bit different because `oc_forward_df` only needs one argument (placename), while `oc_reverse_df` needs two (lat and lon).
- Output options for `oc_forward_df` and `oc_reverse_df` to return all columns, a selection of columns, or to have the selection of columns be bound to the input data frame.
- Also changed output to tibbles

jessesadler commented:
Closing pull request after further changes elsewhere
